File: 2015/php/3/main-part-2.php
<?php

class Position {
    public $x;
    public $y;

    public function __construct($x, $y) {
        $this->x = $x;
        $this->y = $y;
    }

    public function move($direction) {
        switch($direction) {
            case '^':
                $this->y--;
                break;
            case 'v':
                $this->y++;
                break;
            case '<':
                $this->x--;
                break;
            case '>':
                $this->x++;
                break;
        }
    }
}

function main() {
    // get command line arguments
    $args = $_SERVER['argv'];
    if(count($args) < 2) {
        echo "Usage: php main.php <input_file>\n";
        return;
    }

    $filename = $args[1];

    // read input file
    $file = fopen($filename, "r");
    if($file === false) {
        echo "Error: Unable to open file $filename\n";
        return;
    }

    $input = [];
    while(!feof($file)) {
        $line = trim(fgets($file));
        if($line !== "") {
            $input[] = $line;
        }
    }

    fclose($file);

    // create a hashmap to store the number of houses visited
    $houses = [];
    $santa = new Position(0, 0);
    $robo_santa = new Position(0, 0);
    $houses[$santa->x][$santa->y] = 1;
    $houses[$robo_santa->x][$robo_santa->y] += 1;

    $santa_turn = true;
    // loop each line of input
    foreach($input as $line) {
        for($i = 0; $i < strlen($line); $i++) {
            $direction = $line[$i];
            if($santa_turn) {
                $santa->move($direction);
                $houses[$santa->x][$santa->y] = 1;
            } else {
                $robo_santa->move($direction);
                $houses[$robo_santa->x][$robo_santa->y] = 1;
            }
            $santa_turn = !$santa_turn;
        }
    }

    // count the number of houses visited
    $count = 0;
    foreach($houses as $x => $row) {
        $count += count($row);
    }

    echo "Part 2: $count\n";
}

main();

File: 2015/lua/3/main-part-2.lua
local Position = {}

function Position:new()
    o = {}
    setmetatable(o, self)
    self.__index = self
    o:_create()
    return o
end

function Position:_create()
    self.x = 0
    self.y = 0
end

function Position:move(direction)
    if direction == "^" then
        self.y = self.y + 1
    elseif direction == "v" then
        self.y = self.y - 1
    elseif direction == "<" then
        self.x = self.x - 1
    elseif direction == ">" then
        self.x = self.x + 1
    end
end

function main()
    if #arg < 1 then
        print("Usage: lua main.lua <input_file>")
        os.exit(1)
    end

    local filename = arg[1]
    local file = io.open(filename, "r")
    if (file == nil) then
        print("Error: Could not open file " .. filename)
        os.exit(1)
    end

    local content = file:read("*all")
    file:close()

    -- hash table to store the number of houses visited
    local houses = {}

    local santa = Position:new()
    local robo_santa = Position:new()

    local key = santa.x .. "," .. santa.y
    houses[key] = 1
    houses[robo_santa.x .. "," .. robo_santa.y] = houses[robo_santa.x .. "," .. robo_santa.y] + 1
    local santa_turn = true
    -- loop each line
    for line in content:gmatch("[^\r\n]+") do
        -- loop each character
        for c in line:gmatch(".") do
            if santa_turn then
                santa:move(c)
                key = santa.x .. "," .. santa.y
            else
                robo_santa:move(c)
                key = robo_santa.x .. "," .. robo_santa.y
            end

            if houses[key] == nil then
                houses[key] = 1
            else
                houses[key] = houses[key] + 1
            end

            santa_turn = not santa_turn
        end
    end

    local count = 0
    for k, v in pairs(houses) do
        count = count + 1
    end

    print("Part 2: " .. count)
end

main()
